### PR TITLE
[DFSM] Support live updates on login nodes.

### PR DESCRIPTION
**Pull request by gmarciani**
### Description of changes
Add support for live updates on login nodes, using cfn-hup. In particular:
* Move definition of dna.json from login node user data to login node launch template. In this way changes to cluster config file, which are reflected to the dna.json, are reflected on the launch template and so detectable by cfn-hup.
* Add cfn-hup configuration files.
* Add cluster config versioning to the dna.json.
* Add permissions for login nodes to describe the resources in the login fleet nested stack, required by cfn-init and cfn-hup to describe the stack and detect updates.
* Add permissions for login nodes to read/write items in the DDB table, required to put the version of the cluster config file deployed in the nodes.
* Added new unit tests covering IAM permissions for compute nodes, generation of dna.json, user data variables.

Important observations:
1. This PR relies on https://github.com/aws/aws-parallelcluster-cookbook/pull/2640
1. The changes proposed in this PR are equivalent to the ones already approved for the compute fleet in the below PRs:
  1. https://github.com/aws/aws-parallelcluster/pull/6022
1. We are skipping the security exclusions check because in this PR there are temporary files created as part of the template Init section with deterministic names (not good approach). This approach is not introduced in this PR, but is there since the beginning of the product.
1. While adapting the unit test `test_login_nodes_launch_template_properties` for DFSM, we figured out that the unit test was actually wrong as it was validating properties on the compute node launch template rather than liogin nodes launch template. We fixed the test as part of this PR, that's the reason why you would see some apparently unexpected changes in the test.
2. The original unit test covering the generation of dna.json for login nodes has been completely replaced because we moved the definition of dna.json from user data to launch template.

### Tests
* manual create, update, delete checking that login nodes push cluster config version on DDB.
* Integ test `test_update_slurm`
* Integ test `test_cloudwatch_logging`

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/templates/login_nodes_stack.py [hunk 1]
```diff
@@ -27,7 +28,7 @@
     get_user_data_content,
     to_comma_separated_string,
 )
-from pcluster.utils import get_attr, get_http_tokens_setting
+from pcluster.utils import get_attr, get_http_tokens_setting, get_service_endpoint, get_resource_name_from_resource_arn
```
[the remark this answers is not in the document]
Comment (gmarciani) [reply cli/src/pcluster/templates/login_nodes_stack.py]: Addressed

File: cli/tests/pcluster/templates/test_login_nodes_stack.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import pytest
+from assertpy import assert_that
+from freezegun import freeze_time
+
+from pcluster.schemas.cluster_schema import ClusterSchema
+from pcluster.templates.cdk_builder import CDKTemplateBuilder
+from pcluster.utils import load_json_dict, load_yaml_dict
+from tests.pcluster.aws.dummy_aws_api import mock_aws_api
+from tests.pcluster.models.dummy_s3_bucket import dummy_cluster_bucket, mock_bucket_object_utils
+from tests.pcluster.templates.test_cluster_stack import IamPolicyAssertion, get_generated_template_and_cdk_assets
```
[the remark this answers is not in the document]
Comment (gmarciani) [reply cli/tests/pcluster/templates/test_login_nodes_stack.py]: Addressed

File: cli/src/pcluster/resources/login_node/user_data.sh [hunk 3]
```diff
@@ -134,6 +100,12 @@ write_files:
         export HOME="${!HOME_BAK}"
       }
 
+      export PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/opt/aws/bin
+      # Load ParallelCluster environment variables
+      [ -f /etc/profile.d/pcluster.sh ] && . /etc/profile.d/pcluster.sh
+
+      cfn-init -s ${AWS::StackName} -v -c deployFiles -r ${LaunchTemplateResourceId} --region ${AWS::Region} --url ${CloudFormationUrl} --role ${CfnInitRole} || error_exit 'Failed to bootstrap the compute node. Please check /var/log/cfn-init.log in the compute node, or check the cfn-init.log in CloudWatch logs. Please refer to https://docs.aws.amazon.com/parallelcluster/latest/ug/troubleshooting-v3.html#troubleshooting-v3-get-logs for more details on ParallelCluster logs.'
```
Comment (enrico-usai) [inline cli/src/pcluster/resources/login_node/user_data.sh]: Failed to bootstrap the login node.

File: cli/src/pcluster/resources/login_node/user_data.sh [hunk 4]
```diff
@@ -181,17 +153,13 @@ write_files:
...
-        cinc-client --local-mode --config /etc/chef/client.rb --log_level info --force-formatter --no-color --chef-zero-port 8889 --json-attributes /etc/chef/dna.json --override-runlist aws-parallelcluster-entrypoints::config &&
+        cinc-client --local-mode --config /etc/chef/client.rb --log_level info --logfile /var/log/chef-client.log --force-formatter --no-color --chef-zero-port 8889 --json-attributes /etc/chef/dna.json --override-runlist aws-parallelcluster-entrypoints::config &&
         /opt/parallelcluster/scripts/fetch_and_run -postinstall -c /opt/parallelcluster/shared_login_nodes/cluster-config.yaml &&
-        cinc-client --local-mode --config /etc/chef/client.rb --log_level info --force-formatter --no-color --chef-zero-port 8889 --json-attributes /etc/chef/dna.json --override-runlist aws-parallelcluster-entrypoints::finalize &&
+        cinc-client --local-mode --config /etc/chef/client.rb --log_level info --logfile /var/log/chef-client.log --force-formatter --no-color --chef-zero-port 8889 --json-attributes /etc/chef/dna.json --override-runlist aws-parallelcluster-entrypoints::finalize &&
```
Comment (enrico-usai) [inline cli/src/pcluster/resources/login_node/user_data.sh:162]: fix error exit message: `check /var/log/cfn-init.log and /var/log/cloud-init-output.log.`

**Review by enrico-usai — APPROVED** [review]

**Pull request merged**